synapse/storage/databases/main/events.py

```
@@ -1260,6 +1468,135 @@ def _update_current_state_txn(
+                    room_type = event_json.get("content", {}).get(
+                        EventContentFields.ROOM_TYPE
+                    )
+                    sliding_sync_joined_rooms_insert_map["room_type"] = room_type
+                elif event_id == room_encryption_event_id:
+                    encryption_algorithm = event_json.get("content", {}).get(
+                        EventContentFields.ENCRYPTION_ALGORITHM
+                    )
+                    is_encrypted = encryption_algorithm is not None
+                    sliding_sync_joined_rooms_insert_map["is_encrypted"] = is_encrypted
+                elif event_id == room_name_event_id:
+                    room_name = event_json.get("content", {}).get(
+                        EventContentFields.ROOM_NAME
+                    )
+                    sliding_sync_joined_rooms_insert_map["room_name"] = room_name
+                else:
+                    raise AssertionError(
+                        f"Unexpected event_id (we should not be fetching extra events): {event_id}"
+                    )
+
+            # Update the `sliding_sync_joined_rooms` table
+            args: List[Any] = [
+                room_id,
+                # Even though `Mapping`/`Dict` have no guaranteed order, some
+                # implementations may preserve insertion order so we're just going to
+                # choose the best possible answer by using the "first" event ID which we
+                # will assume will have the greatest `stream_ordering`. We really just
+                # need *some* answer in case we are the first ones inserting into the
+                # table and in reality,
+                # `_update_sliding_sync_tables_with_new_persisted_events_txn()` is run
+                # after this function to update it to the correct latest value. This is
+                # just to account for things changing in the future.
+                next(iter(to_insert.values())),
```

When `to_insert` is empty, this raises a `StopIteration` exception/error that is completely swallowed by the transaction and silenty errors (never logs anything went wrong).

Why does `raise StopIteration()` not get caught by the [database trasnsaction `except Exception` handler](https://github.com/element-hq/synapse/blob/a308d99f30d7e660115e355c54c37ac149cdbe53/synapse/storage/database.py#L938-L941) but `raise AssertionError("fake error")` does? They are both `Exception`

---

Seems related to https://peps.python.org/pep-0479/ but that makes it seem like with that change (since Python 3.5), `StopIteration` shouldn't be silently ignored like it is now.

I'm guessing this is a bad interaction between the `with` context managers at various points and `make_deferred_yieldable(...)` swallowing it.

There have been more recent related issues around this like https://github.com/python/cpython/issues/112182

---

Created https://github.com/element-hq/synapse/issues/17597 to track this